packages/common/infra/src/op/consumer.ts

```diff
@@ -0,0 +1,188 @@
+            { transfer: transferables }
+          );
+        },
+        error: error => {
+          this.port.postMessage({
+            type: 'return',
+            id: msg.id,
+            error: error as Error,
+          } satisfies ReturnMessage);
+        },
+        complete: () => {
+          this.processing.delete(msg.id);
+        },
+      });
+  };
+
+  private readonly handleSubscribeMessage: MessageHandlers['subscribe'] =
+    msg => {
+      const abortController = new AbortController();
+      this.processing.set(msg.id, abortController);
+
+      this.ob$(msg, abortController.signal).subscribe({
+        next: data => {
+          const transferables = fetchTransferables(data);
+          this.port.postMessage(
+            {
+              type: 'next',
+              id: msg.id,
+              data,
+            } satisfies SubscriptionNextMessage,
+            { transfer: transferables }
+          );
+        },
+        error: error => {
+          this.port.postMessage({
+            type: 'error',
+            id: msg.id,
+            error: error as Error,
+          } satisfies SubscriptionErrorMessage);
+        },
+        complete: () => {
+          this.port.postMessage({
+            type: 'complete',
+            id: msg.id,
+          } satisfies SubscriptionCompleteMessage);
+          this.processing.delete(msg.id);
+        },
+      });
+    };
+
+  private readonly handleCancelMessage: MessageHandlers['cancel'] &
+    MessageHandlers['unsubscribe'] = msg => {
+    const abortController = this.processing.get(msg.id);
+    if (!abortController) {
+      return;
+    }
+
+    abortController.abort();
+  };
+
+  register<Op extends OpNames<Ops>>(op: Op, handler: OpHandler<Ops, Op>) {
+    this.registeredOpHandlers.set(op, handler);
+  }
+
+  before<Op extends OpNames<Ops>>(
+    op: Op,
+    handler: (...input: OpInput<Ops, Op>) => void
+  ) {
+    this.eventBus.on(`before:${op}`, handler);
+  }
+
+  after<Op extends OpNames<Ops>>(
+    op: Op,
+    handler: (...args: [...OpInput<Ops, Op>, OpOutput<Ops, Op>]) => void
+  ) {
+    this.eventBus.on(`after:${op}`, handler);
+  }
+
+  /**
+   * @internal
+   */
+  ob$(op: CallMessage | SubscribeMessage, signal: AbortSignal) {
+    return defer(() => {
+      const handler = this.registeredOpHandlers.get(op.name as any);
+      if (!handler) {
+        throw new Error(
+          `Handler for operation [${op.name}] is not registered.`
+        );
+      }
+
+      const ret$ = handler(op.payload, { signal });
+
+      let ob$: Observable<any>;
+      if (ret$ instanceof Promise) {
+        ob$ = from(ret$);
+      } else if (ret$ instanceof Observable) {
+        ob$ = ret$;
+      } else {
+        ob$ = of(ret$);
+      }
+
+      return ob$.pipe(share(), takeUntil(fromEvent(signal, 'abort')));
+    });
+  }
+
+  destroy() {
+    super.close();
+    this.registeredOpHandlers.clear();
+    this.processing.forEach(controller => {
+      controller.abort();
```

Comment:
maybe add abort reason for auto recover

```suggestion
      controller.abort("DESTROY");
```

Reply:
what can be recovered